src/Action/AppendFormFieldElementAction.php

```diff
@@ -39,30 +36,70 @@ final class AppendFormFieldElementAction
      */
     private $twig;
 
-    public function __construct(Environment $twig, Pool $pool, AdminHelper $helper)
+    /**
+     * @var AdminFetcherInterface
+     */
+    private $adminFetcher;
+
+    /**
+     * NEXT_MAJOR: Restrict second param to AdminFetcherInterface.
+     *
+     * @param Pool|AdminFetcherInterface $poolOrAdminFetcher
+     */
+    public function __construct(Environment $twig, $poolOrAdminFetcher, AdminHelper $helper)
     {
-        $this->pool = $pool;
         $this->helper = $helper;
         $this->twig = $twig;
+
+        if ($poolOrAdminFetcher instanceof AdminFetcherInterface) {
+            $this->adminFetcher = $poolOrAdminFetcher;
+        } elseif ($poolOrAdminFetcher instanceof Pool) {
+            @trigger_error(sprintf(
+                'Passing other type than %s in argument 2 to %s() is deprecated since sonata-project/admin-bundle 3.x'
+                .' and will throw %s exception in 4.0.',
+                AdminFetcherInterface::class,
+                __METHOD__,
+                \TypeError::class
+            ), \E_USER_DEPRECATED);
+
+            $this->adminFetcher = new AdminFetcher($poolOrAdminFetcher);
+        } else {
+            throw new \TypeError(sprintf(
+                'Argument 2 passed to "%s()" must be either an instance of %s or %s, %s given.',
+                __METHOD__,
+                Pool::class,
+                AdminFetcherInterface::class,
+                \is_object($poolOrAdminFetcher) ? 'instance of "'.\get_class($poolOrAdminFetcher).'"' : '"'.\gettype($poolOrAdminFetcher).'"'
+            ));
+        }
     }
 
     /**
      * @throws NotFoundHttpException
      */
     public function __invoke(Request $request): Response
     {
-        $code = $request->get('code');
-        $elementId = $request->get('elementId');
-        $objectId = $request->get('objectId');
-        $uniqid = $request->get('uniqid');
-
-        $admin = $this->pool->getInstance($code);
-        $admin->setRequest($request);
+        // NEXT_MAJOR: Remove this BC-layer.
+        if (null === $request->get('_sonata_admin')) {
+            @trigger_error(
+                'Not passing "_sonata_admin" value in the request is deprecated since sonata-project/admin-bundle 3.x'
+                .' and will throw %s exception in 4.0.',
+                \E_USER_DEPRECATED
+            );
+
+            $request->request->set('_sonata_admin', $request->get('code'));
```

Comment:
```suggestion
            $request->query->set('_sonata_admin', $request->get('code'));
```

setting the parameter to the query string looks better IMO (In practice it doesnt' matter because is fetched with `Request::get()`), in the future I think it would be nice to not use `Request::get()` and use the proper one.

Reply:
Fixed